chrysom/client.go

```diff
@@ -129,140 +162,229 @@ func validateConfig(config *ClientConfig) error {
 	}
 	return nil
 }
+func shouldUseJWTAcquirer(options acquire.RemoteBearerTokenAcquirerOptions) bool {
+	return len(options.AuthURL) > 0 && options.Buffer != 0 && options.Timeout != 0
+}
 
-func determineTokenAcquirer(config ClientConfig) (acquire.Acquirer, error) {
-	defaultAcquirer := &acquire.DefaultAcquirer{}
-	if config.Auth.JWT.AuthURL != "" && config.Auth.JWT.Buffer != 0 && config.Auth.JWT.Timeout != 0 {
-		return acquire.NewRemoteBearerTokenAcquirer(config.Auth.JWT)
+func buildTokenAcquirer(auth *Auth) (acquire.Acquirer, error) {
+	if shouldUseJWTAcquirer(auth.JWT) {
+		return acquire.NewRemoteBearerTokenAcquirer(auth.JWT)
+	} else if len(auth.Basic) > 0 {
+		return acquire.NewFixedAuthAcquirer(auth.Basic)
 	}
+	return &acquire.DefaultAcquirer{}, nil
+}
 
-	if config.Auth.Basic != "" {
-		return acquire.NewFixedAuthAcquirer(config.Auth.Basic)
+func validateGetItemsInput(input *GetItemsInput) error {
+	if input == nil {
+		return ErrUndefinedInput
 	}
 
-	return defaultAcquirer, nil
+	if len(input.Bucket) < 1 {
+		return ErrBucketEmpty
+	}
+	return nil
 }
 
-func (c *Client) GetItems(owner string) ([]model.Item, error) {
-	request, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/store/%s", c.remoteStoreAddress, c.bucketName), nil)
+func (c *Client) makeRequest(owner, method, URL string, body io.Reader) (*http.Request, error) {
+	r, err := http.NewRequest(method, URL, body)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrNewRequestFailure, err.Error())
 	}
-	err = acquire.AddAuth(request, c.auth)
+	err = acquire.AddAuth(r, c.auth)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrAuthAcquirerFailure, err.Error())
+	}
+
+	if len(owner) > 0 {
+		r.Header.Set(store.ItemOwnerHeaderKey, owner)
+	}
+
+	return r, nil
+}
+
+func (c *Client) do(r *http.Request) (*doResponse, error) {
+	resp, err := c.client.Do(r)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrDoRequestFailure, err.Error())
 	}
+	defer resp.Body.Close()
 
-	request.Header.Set(store.ItemOwnerHeaderKey, owner)
+	var dResp = doResponse{
+		Code: resp.StatusCode,
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return &dResp, fmt.Errorf("%w: %s", ErrReadingBodyFailure, err.Error())
+	}
+	dResp.Body = bodyBytes
+	return &dResp, nil
+}
 
-	response, err := c.client.Do(request)
+// GetItems fetches all items in a bucket that belong to a given owner.
+func (c *Client) GetItems(input *GetItemsInput) (*GetItemsOutput, error) {
```

Comment:
`*GetItemsOutput` is a pointer to a struct that only has a slice, which is a pointer.   Why can't we just return the slice?

Reply:
I have the same question about other `Output` structs.

Reply:
Responded in comment above.

Reply:
Passing by value seems better in this instance.

I would second what @kristinaspring says above.  I'm not so much worried about speed of a particular call.  Over time and at scale, the garbage collector performance tends to dominate.  Passing by value removes the garbage collector from that part of the callstack, at least insofar as the parameters are concerned.

It's best to make the decision to pass by value or pointer based on API design rather than performance.  If something is immutable or shouldn't be modified by a function, passing by value is a great way to communicate that without having to put a lot of verbiage in `godoc`.  Alternatively, if the function does or may modify something, passing by pointer communicates that succinctly as well.

Obviously, if you benchmark and find significant performance gains by switching to pointers, by all means pass by pointer.  But 4000ns/op, or in other words 4 microseconds/op, doesn't seem all that huge.